test/integration_tests/prototype/hifi_gan_pipeline_test.py

```diff
@@ -0,0 +1,65 @@
+import os
+import sys
+
+import pytest
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", "..", "examples"))
+import torch
+import torchaudio
+import torchaudio.functional as F
+from source_separation.utils.metrics import sdr
+from torchaudio.prototype.pipelines import HIFIGAN_GENERATOR_V3_LJSPEECH
+
+from torchaudio.transforms import MelSpectrogram
+
+
+def test_smoke_hifi_gan_bundle():
+    """Smoke test of downloading weights for pretraining model"""
+    HIFIGAN_GENERATOR_V3_LJSPEECH.get_vocoder()
+
+
+@pytest.mark.parametrize(
+    "lang",
+    [
+        ("en"),
+        ("en2"),
+    ],
+)
+def test_hifi_gan_pretrained_weights(lang, sample_speech):
+    """Test that HiFiGAN bundle can reconstruct waveform from mel spectrogram with sufficient SDR score"""
+    bundle = HIFIGAN_GENERATOR_V3_LJSPEECH
+    # Config params from
+    # https://github.com/jik876/hifi-gan/blob/4769534d45265d52a904b850da5a622601885777/config_v3.json#L17-L27
+    hop_size = 256
+    n_fft = 1024
+    win_length = 1024
+    f_min, f_max = 0, 8000
+    ref_waveform_raw, ref_sample_rate = torchaudio.load(sample_speech)
+    ref_waveform = F.resample(ref_waveform_raw, orig_freq=ref_sample_rate, new_freq=bundle.sample_rate)
+    # Generate mel spectrogram in the same way as the original HiFiGAN implementation does
+    # https://github.com/jik876/hifi-gan/blob/4769534d45265d52a904b850da5a622601885777/meldataset.py#L49-L72
+    mel_transform = MelSpectrogram(
+        sample_rate=bundle.sample_rate,
+        n_fft=n_fft,
+        win_length=win_length,
+        hop_length=hop_size,
+        f_min=f_min,
+        f_max=f_max,
+        n_mels=bundle._params["in_channels"],
+        normalized=False,
+        mel_scale="slaney",
+        norm="slaney",
+        center=True,
+    )
+    mel_spectrogram = mel_transform(ref_waveform)
+    mel_spectrogram = torch.log(torch.clamp(mel_spectrogram, min=1e-5))
```

Comment:
The design principle of Bundle object is to free users from re-implementing the expected preprocessing like this. 

Please checkout the followings for the detail

- https://pytorch.org/audio/stable/pipelines.html#torchaudio-pipelines
- https://pytorch.org/audio/stable/tutorials/tacotron2_pipeline_tutorial.html#phoneme-based-encoding

Can you move the construction of preprocessing pipeline into bundle?

Reply:
Sounds good, I've added:
 - a class `HiFiGANMelSpectrogram` performing waveform-to-mel transform in the same way as original HiFiGAN code
 - a factory function `hifigan_mel_spectrogram`
 - `HiFiGANGeneratorBundle.get_mel_transform` method, which returns `HiFiGANMelSpectrogram`
 - a unit test which checks that `HiFiGANMelSpectrogram` is indeed equivalent to the original pipeline

That said, the integration test for reconstructing waveform from a real audio is not properly working yet - even the shift-invariant metric @nateanl suggested gives negative SNR. The reconstructed audio has a noticeable noise, maybe that's the reason - I'll keep looking into this

Reply:
> Sounds good, I've added:
> 
> * a class `HiFiGANMelSpectrogram` performing waveform-to-mel transform in the same way as original HiFiGAN code
> * a factory function `hifigan_mel_spectrogram`
> * `HiFiGANGeneratorBundle.get_mel_transform` method, which returns `HiFiGANMelSpectrogram`
> * a unit test which checks that `HiFiGANMelSpectrogram` is indeed equivalent to the original pipeline
> 

Another thing about the design around the Bundle is that we want to make it convenient but we want to limit the number of public interfaces we introduce.

In PyTorch, we make it rule to take 2 release cycles for BC-breaking changes to stable public interfaces. That's like 6+ month, which is a long time. (prototypes are okay to break backward compatibility anytime) So any public interface is basically maintenance cost since right when it is introduced.

For this reason, in Bundle and pipelines, the classes that implement pre/post-processing, (or even model structures themself) are not public interface, but their calling conventions (the signatures of `forward` method) are. The design is centered around the usage of pre-trained models, and we take advantage of the fact that Python is Duck Typing language here.

So, it is important to make pre-processings accessible for users, but pre-processings implementation themself are not important enough for us to declare them as public. Otherwise, we have to introduce ton of preprocessing classes as public interface, which is not maintainable.

An fine example is [source separation bundle](https://pytorch.org/audio/stable/pipelines.html#source-separation). [CONVTASNET_BASE_LIBRI2MIX](https://pytorch.org/audio/stable/generated/torchaudio.pipelines.CONVTASNET_BASE_LIBRI2MIX.html#torchaudio.pipelines.CONVTASNET_BASE_LIBRI2MIX) and [HDEMUCS_HIGH_MUSDB_PLUS](https://pytorch.org/audio/stable/generated/torchaudio.pipelines.HDEMUCS_HIGH_MUSDB_PLUS.html#torchaudio.pipelines.HDEMUCS_HIGH_MUSDB_PLUS) share the same usage and interface but the returned models are completely different as they process different kind of audios.

For this reason,

- We do not need a factory function `hifigan_mel_spectrogram`.
- I am inclined to call `HiFiGANMelSpectrogram` to something a bit more general, like preprocessing, but in the context of GAN training preprocessing is too ambiguous so I am open to other ideas.

Reply:
wait, actually, I might be misunderstanding an important thing. What's the expected usecase/user-flow for this bundle? For the actual vocoding, they are connected with some *other* mechanism that generates melspectrogram, right, and is such mechanism going to be added to this bundle??

Reply:
For the complete flow we would need an external text-to-melspectrogram mechanism - for example, the one from Tacotron2. This allows a text -> melspectrogram -> waveform flow. Here is the example snippet: https://github.com/pytorch/audio/pull/2921/files#diff-12a2fb327badd19aa9ee81293aa2815ae0249fbb3cfada531e5552af6d02dd33R43-R66

In this case do we need to add Tacotron2 text-to-melspectrogram component to the HiFiGAN bundle? If we want to use HiFiGAN with different  text-to-melspectrogram components, would we define a bundle for each combination? I see that this is indeed how Tacotron2 bundles are organized https://github.com/pytorch/audio/blob/f0986b6044c43ac2cb17afb42352fadd43c5deb2/torchaudio/pipelines/_tts/impl.py#L187-L203